gpt4all-chat/src/database.cpp

```diff
@@ -2036,20 +2078,222 @@ QList<int> Database::searchEmbeddings(const std::vector<float> &query, const QLi
+        qWarning() << "Database ERROR: Failed to exec embeddings query:" << q.lastError();
+        return {};
+    }
+    return searchEmbeddingsHelper(query, q, nNeighbors);
+}
+
+QList<int> Database::scoreChunks(const std::vector<float> &query, const QList<int> &chunks)
+{
+    QList<QString> chunkStrings;
+    for (int id : chunks)
+        chunkStrings << QString::number(id);
+    QSqlQuery q(m_db);
+    if (!q.exec(GET_CHUNK_EMBEDDINGS_SQL.arg(chunkStrings.join("', '")))) {
+        qWarning() << "Database ERROR: Failed to exec embeddings query:" << q.lastError();
+        return {};
+    }
+    return searchEmbeddingsHelper(query, q, chunks.size());
+}
+
+QList<Database::BM25Query> Database::queriesForFTS5(const QString &input)
+{
+    // Escape double quotes by adding a second double quote
+    QString escapedInput = input;
+    escapedInput.replace("\"", "\"\"");
+
+    static QRegularExpression spaces("\\s+");
+    QStringList oWords = escapedInput.split(spaces, Qt::SkipEmptyParts);
+
+    QList<BM25Query> queries;
+
+    // Start by trying to match the entire input
+    BM25Query e;
+    e.type = 0; // exact
+    e.input = oWords.join(" ");
+    e.query = "\"" + oWords.join(" ") + "\"";
+    e.qlength = oWords.size();
+    e.ilength = oWords.size();
+    queries << e;
+
+    // https://github.com/igorbrigadir/stopwords?tab=readme-ov-file
+    // Lucene, Solr, Elastisearch
+    static const QList<QString> stopWords = {
+        "a", "an", "and", "are", "as", "at", "be", "but", "by",
+        "for", "if", "in", "into", "is", "it", "no", "not", "of",
+        "on", "or", "such", "that", "the", "their", "then", "there",
+        "these", "they", "this", "to", "was", "will", "with"
+    };
+
+    QStringList rWords;
+    for (const QString &w : oWords)
+        if (!stopWords.contains(w.toLower()))
+            rWords << w;
+
+    QStringList quotedWords;
+    for (const QString &w : rWords)
+        quotedWords << "\"" + w + "\"";
+
+    BM25Query b;
+    b.type = 1; // broad search
+    b.input = oWords.join(" ");
+    b.query = "(" + quotedWords.join(" OR ") + ")";
+    b.qlength = 1; // length of phrase
+    b.ilength = oWords.size();
+    b.rlength = oWords.size() - rWords.size();
+    queries << b;
+    return queries;
+}
+
+QList<int> Database::searchBM25(const QString &query, const QList<QString> &collections, BM25Query &bm25q, int k)
+{
+    struct SearchResult { int chunkId; float dist; };
+    QList<BM25Query> bm25Queries = queriesForFTS5(query);
+
+    QList<SearchResult> results;
+    for (auto bm25Query : bm25Queries) {
+        QSqlQuery sqlQuery(m_db);
+        sqlQuery.prepare(SELECT_CHUNKS_FTS_SQL.arg(k));
+        sqlQuery.addBindValue(bm25Query.query);
+
+        if (!sqlQuery.exec()) {
+            qWarning() << "Database ERROR: Failed to execute BM25 query:" << sqlQuery.lastError();
+            return {};
+        }
+
+        if (!sqlQuery.next())
+            continue;
+
+        bm25q = bm25Query;
```

Comment:
Should add a brief comment explaining what this represents, e.g. "return the query used to produce the results".